### PR TITLE
Upgrade for v3

### PR DESCRIPTION
**Pull request by ptuchik**
Tested on my project, fully working as with v2.x

### CODE REVIEW

File: src/Message/UpdatePaymentMethodRequest.php [hunk 1]
```diff
@@ -28,11 +27,12 @@ public function getData()
      * Send the request with specified data
      *
      * @param  mixed $data The data to send
+     *
      * @return ResponseInterface
      */
     public function sendData($data)
     {
-        $response = $this->braintree->paymentMethod()->update($data['token'], $data['options']);
+        $response = $this->braintree->paymentMethod()->update($data['token'], $data);
```
Comment (barryvdh) [inline src/Message/UpdatePaymentMethodRequest.php]: Is this change related to the v3 upgrade?

File: src/Message/FindCustomerRequest.php [hunk 2]
```diff
@@ -16,12 +19,18 @@ public function getData()
...
+     * @return \Omnipay\Braintree\Message\CustomerResponse|\Omnipay\Common\Message\ResponseInterface
+     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
      */
     public function sendData($data)
     {
-        $response = $this->braintree->customer()->find($this->getCustomerId());
+        try {
+            $response = $this->braintree->customer()->find($this->getCustomerId());
+        } catch (NotFound $exception) {
+            throw new NotFoundHttpException($exception->getMessage());
```
Comment (barryvdh) [inline src/Message/FindCustomerRequest.php]: I wouldn't throw this error, but rather `Omnipay\Common\Exception\InvalidResponseException`. This error could cause the application to render a 404 page instead of an error page (for example Laravel converts the HttpExceptions to specific errors)

File: src/Message/UpdatePaymentMethodRequest.php (hunk 1 above)
Comment (ptuchik) [reply src/Message/UpdatePaymentMethodRequest.php]: No, this was a bug in previous version. I was fixed there on my fork when using with v2.x and it is still actual.

File: src/Message/FindCustomerRequest.php (hunk 2 above)
Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: Thanks for your comment. I agree with you, but the main reason I'm throwing not found exception is for those (like me), who wants to recreate payment profile if the existing one is invalid or not found, without showing the error. So I'm throwing the general not found here, catching it in my Billable trait and recreating the profile. And I used generic Not Found instead of Braintree's one to have the logic working with other gateways also.

Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: I think it's better then to create your own exception that extends the omnipay invalidresponse exception.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: I agree, the problem is that if I create may own in this package, I can't catch it generalized, because other payment gateway packages will not have that exception. That's why I used Symphony's one

File: src/Message/UpdatePaymentMethodRequest.php (hunk 1 above)
Comment (ptuchik) [reply src/Message/UpdatePaymentMethodRequest.php]: This can be changed to `$data['options'] ?? []` as well

Comment (ptuchik) [reply src/Message/UpdatePaymentMethodRequest.php]: I just tested with `$data['options'] ?? []` instead of `data`, it is not working because of `Braintree\Util`'s validation. So the only way it will work is the one I've commited

File: src/Message/FindCustomerRequest.php (hunk 2 above)
Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: If you extend the InvalidResponse one, all gateway can catch that. You can just add specific logic for that exception.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: Yes, they can, but they can’t detect that it is Not Found exception, they will catch as a just Invalid Response

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: Anyways if I will just remove that part, and leave as is, it will throw Braintree's exception, like in all other places, instead of Omnipay's Invalid Response. So I've only generalized this, to have Not Found behave like Symphony's Not Found

Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: Correct, but if they want to implement specific behavior for Braintree, they can also use the Exception.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: Agreed, but in that case they have to implement in their code something like `catch Braintree/NotFound $exception` and then `catch AnotherGateway\NotFound $exception` etc.. so I wrote it here to give ability for developers to catch just one `NotFoundHttpException` for all gateways they may have

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: The best way would be if Omnipay would have it's own NotFoundException extended from InvalidResponseException, and all gateways could through it in such cases, to differentiate all Invalid exceptions from NotFound

Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: We could do that perhaps. You can create a PR on omnipay/common if you want.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: OK, thank you, let me do it just now please. And update this one also

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: Created, here is the PR for Omnipay Common https://github.com/thephpleague/omnipay-common/pull/187 and I've updated this one as well

Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: Can you remove the try/catch for now and leave it as before, so we can merge this? We can update it once we merge the PR and a new tag is released.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: OK, just 2 questions before I remove:
1. When you think the tag will be released?
2. May I leave the previous version with Symphony's NotFound until the release of Common package? To be able to handle it without hardcoding "Braintree" namespace in my application

Comment (barryvdh) [reply src/Message/FindCustomerRequest.php]: 1. probably with a week or so, but not sure.
2. No but you could use a gateway specific exception.

Comment (ptuchik) [reply src/Message/FindCustomerRequest.php]: OK, done

**Comment by ptuchik** [comment]
Thank you

**Comment by ptuchik** [comment]
Could you please let me know, when this merge will be available to Packagist?

**Comment by barryvdh** [comment]
Please test this using version `3.x@dev`, which should be published to Packagist now.

**Comment by barryvdh** [comment]
Only thing blocking is this error: https://travis-ci.org/thephpleague/omnipay-braintree/jobs/383745948

**Comment by ptuchik** [comment]
I see, maybe it needs to be `number_format`-ed with currency specific format before passing to validator?

**Comment by barryvdh** [comment]
Not sure, I marked it as skipped. Don't think that needs to be intended behavior.
